packages/langium/src/documents/document-builder.ts

```diff
@@ -4,23 +4,19 @@
  * terms of the MIT License, which is available in the project root.
  ******************************************************************************/
 
-import { Connection, Diagnostic } from 'vscode-languageserver/node';
+import { CancellationToken, Connection, Diagnostic } from 'vscode-languageserver';
 import { DocumentValidator } from '../validation/document-validator';
 import { LangiumParser, ParseResult } from '../parser/langium-parser';
 import { ScopeComputation } from '../references/scope';
 import { LangiumServices } from '../services';
-import { LangiumDocument, LangiumDocuments } from './document';
+import { DocumentState, LangiumDocument, LangiumDocuments } from './document';
 import { IndexManager } from '../index/index-manager';
 import { URI } from 'vscode-uri';
+import { interruptAndCheck, MaybePromise } from '../utils/promise-util';
+import { Linker } from '../references/linker';
 
 export interface DocumentBuilder {
-    build(document: LangiumDocument): BuildResult
-    /**
-     * Validates a document and creates diagnostic information.
-    * @param document to be validated
-    * @returns Diagnostic array or undefined if no validation issues found
-    */
-    validate(document: LangiumDocument): Diagnostic[] | undefined
+    build(document: LangiumDocument, cancelToken?: CancellationToken): Promise<BuildResult>
```

Comment:
Please add some documentation to the public API methods, including a `@throws OperationCanceled` with some explanation.